bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogManagerForEntryLogPerLedger.java

```diff
@@ -0,0 +1,463 @@
+         * entrylogperledger scenario, since entries of a ledger go
+         * to a entrylog (even during compaction) and SyncThread
+         * drives periodic checkpoint logic.
+         */
+
+    }
+
+    @Override
+    public void prepareEntryMemTableFlush() {
+        // do nothing
+    }
+
+    @Override
+    public boolean commitEntryMemTableFlush() throws IOException {
+        // lock it only if there is new data
+        // so that cache accesstime is not changed
+        Set<BufferedLogChannel> copyOfCurrentLogs = getCopyOfCurrentLogs();
+        for (BufferedLogChannel currentLog : copyOfCurrentLogs) {
+            if (reachEntryLogLimit(currentLog, 0L)) {
+                Long ledgerId = currentLog.getLedgerIdAssigned();
+                acquireLock(ledgerId);
+                try {
+                    if (reachEntryLogLimit(currentLog, 0L)) {
+                        log.info("Rolling entry logger since it reached size limitation for ledger: {}", ledgerId);
+                        createNewLog(ledgerId);
+                    }
+                } finally {
+                    releaseLock(ledgerId);
+                }
+            }
+        }
+        /*
+         * in the case of entrylogperledger, SyncThread drives
+         * checkpoint logic for every flushInterval. So
+         * EntryMemtable doesn't need to call checkpoint in the case
+         * of entrylogperledger.
+         */
+        return false;
+    }
+
+    /*
+     * this is for testing purpose only. guava's cache doesnt cleanup
+     * completely (including calling expiry removal listener) automatically
+     * when access timeout elapses.
+     *
+     * https://google.github.io/guava/releases/19.0/api/docs/com/google/
+     * common/cache/CacheBuilder.html
+     *
+     * If expireAfterWrite or expireAfterAccess is requested entries may be
+     * evicted on each cache modification, on occasional cache accesses, or
+     * on calls to Cache.cleanUp(). Expired entries may be counted by
+     * Cache.size(), but will never be visible to read or write operations.
+     *
+     * Certain cache configurations will result in the accrual of periodic
+     * maintenance tasks which will be performed during write operations, or
+     * during occasional read operations in the absence of writes. The
+     * Cache.cleanUp() method of the returned cache will also perform
+     * maintenance, but calling it should not be necessary with a high
+     * throughput cache. Only caches built with removalListener,
+     * expireAfterWrite, expireAfterAccess, weakKeys, weakValues, or
+     * softValues perform periodic maintenance.
+     */
+    @VisibleForTesting
+    void doEntryLogMapCleanup() {
+        ledgerIdEntryLogMap.cleanUp();
+    }
+
+    @VisibleForTesting
+    ConcurrentMap<Long, EntryLogAndLockTuple> getCacheAsMap() {
+        return ledgerIdEntryLogMap.asMap();
+    }
+    /*
+     * Returns writable ledger dir with least number of current active
```

Comment:
This way if you lose a disk you will partially lose a ledger.
I wonder if it would be better to put all data of a ledger in the same directory.

In the future maybe we can think about selecting directory with an algorithm which takes into account some writeflag/metadata. Thinking about multi tenancy with a limited number of bookies

Certainly if you have few ledgers and many directory with this implementation you can use all of the disk space.



@jvrao

Reply:
I'm not sure if it is correct to expect to pick same ledgerDir for a ledger. If a ledgerDir is full then it doesn't makes sense to pick the same ledgerdir when there is other available ledgerdirs.

This is not any new behavior I'm introducing here, it is the existing behavior. Here I'm making sure I'm picking ledgerDir with least number of current active logs, so no ledgerdir is overloaded.